### DISCUSSION TITLE
Will this library work on vanilla Javascript / Jquery?

### DISCUSSION THREAD

**Discussion by AmeyDeshTQ** [dolanmiu/docx#2130] [Q&A] [answered]
I'm able to generate documents using vanilla javascript  and a simple HTML file  (not hosted on the server) however I'm facing issues when reading existing document mentioned in the below example 

https://github.com/dolanmiu/docx/blob/master/demo/85-template-document.ts

Here's a screenshot of error 

![image](https://github.com/dolanmiu/docx/assets/126790075/6d351907-3edd-44f4-ba45-c6389b9efcd4)

Here's a link to code-sandbox - https://codesandbox.io/s/zealous-elion-7dt2hv?file=/src/index.js

When I click on 'Read Existing Document'  I get those errors 

Question is can we use this library without Node.js and without hosting it on the server and just use javascript/jquery?

**Comment by dolanmiu** [comment, marked as answer]
It does yes, please refer to:

https://github.com/dolanmiu/docx/blob/master/demo/browser-demo.html

**Reply by AmeyDeshTQ** [reply]
Yes, it works while generating the Word documents however when I try to open an existing document it's not working

**Reply by dolanmiu** [reply]
`docx` does not support opening documents. only modifying documents or creating them

**Reply by AmeyDeshTQ** [reply]
I mean modifying an existing document? is this not possible with Javascript? 

I read this is possible in one of the other discussions  - 
https://github.com/dolanmiu/docx/discussions/2109

**Reply by dolanmiu** [reply]
It is possible yes, look at this demo here:

https://github.com/dolanmiu/docx/blob/master/demo/85-template-document.ts

**Reply by AmeyDeshTQ** [reply]
But we need Node for this isn't and not possible within the browser?

**Reply by dolanmiu** [reply]
It is possible in the browser.

It's just that in this example, it uses `fs` to load the word document in and to load images. If you find an alternative browser way to do so, then its possible yes

`patchDocument` takes one of the following:

```
Buffer | string | number[] | Uint8Array | ArrayBuffer | Blob | NodeJS.ReadableStream
```

**Reply by AmeyDeshTQ** [reply]
thanks, I've taken out the fs to load, instead I'm reading base64 doc to a blob first and then passing it to patch document, here's the screenshot below

![image](https://github.com/dolanmiu/docx/assets/126790075/d99bdfeb-9696-4503-a8cb-fdcefb79b5ba)

**Reply by AmeyDeshTQ** [reply]
@dolanmiu - I'm still getting the below error 

![image](https://github.com/dolanmiu/docx/assets/126790075/9f7dc771-5ad8-4ccc-9baf-5deac875d53d)


I've used Webpack and here's my package.json file

{
  "scripts": {
    "build": "webpack",
    "watch": "webpack --watch"
  },
  "dependencies": {
    "buffer": "^6.0.3",
    "docx": "^8.0.4",
    "file-saver": "^2.0.5"
  },
  "devDependencies": {
    "webpack": "^5.83.1",
    "webpack-cli": "^5.1.1"
  }
}

**Reply by dolanmiu** [reply]
Because NodeBuffer is not part of the browser (because its not node)

**Comment by AmeyDeshTQ** [comment]
@dolanmiu - Ok I managed to resolve the node buffer no defined issue by adding node poly fill plugin  for webpack - https://github.com/Richienb/node-polyfill-webpack-plugin  however I'm now getting the below issue 




![image](https://github.com/dolanmiu/docx/assets/126790075/013c2271-9966-4e14-8c6d-f73e5677d305)
 

This error is thrown in the Packers module - here are some screenshots 

**Code**
![image](https://github.com/dolanmiu/docx/assets/126790075/9f509c7b-5662-4c01-8542-bc8dd8115f17)

**Browser**

![image](https://github.com/dolanmiu/docx/assets/126790075/cc7d4aec-592d-44b1-a076-83f785ab3102)


PS: The same code runs under Node.js though

**Reply by AmeyDeshTQ** [reply]
@dolanmiu  - It worked by not using the packer and just getting the base64 from the doc buffer 

.then((doc) => {
       
        base64OutputDoc = doc.toString('base64');
        console.log('converted to base64');
});

**Reply by dolanmiu** [reply]
Nice, glad you got it working
